presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+
+            checkExpressionTypeMatching(node.getAssignments());
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            ListMultimap<Symbol, Symbol> symbolMapping = node.getSymbolMapping();
+            for (Symbol keySymbol : symbolMapping.keySet()) {
+                List<Symbol> valueSymbols = symbolMapping.get(keySymbol);
+                Type expectedType = types.get(keySymbol);
+                for (Symbol valueSymbol : valueSymbols) {
+                    verifyTypeSignature(keySymbol, expectedType.getTypeSignature(), types.get(valueSymbol).getTypeSignature());
+                }
+            }
+
+            return null;
+        }
+
+        private void checkSignatureTypeMatching(Map<Symbol, Signature> functions)
+        {
+            for (Map.Entry<Symbol, Signature> entry : functions.entrySet()) {
+                TypeSignature expectedTypeSignature = types.get(entry.getKey()).getTypeSignature();
+                TypeSignature actualTypeSignature = entry.getValue().getReturnType();
+                verifyTypeSignature(entry.getKey(), expectedTypeSignature, actualTypeSignature);
+            }
+        }
+
+        private void checkFunctionCallTypeMatching(Map<Symbol, FunctionCall> functionCalls)
+        {
+            for (Map.Entry<Symbol, FunctionCall> entry : functionCalls.entrySet()) {
+                Type expectedType = types.get(entry.getKey());
+                Type actualType = getExpressionTypes(session, metadata, sqlParser, types, entry.getValue()).get(entry.getValue());
+                verifyTypeSignature(entry.getKey(), expectedType.getTypeSignature(), actualType.getTypeSignature());
+            }
+        }
+
+        private void checkExpressionTypeMatching(Map<Symbol, Expression> expressions)
+        {
+            for (Map.Entry<Symbol, Expression> entry : expressions.entrySet()) {
+                Type expectedType = types.get(entry.getKey());
+                if (entry.getValue() instanceof QualifiedNameReference) {
+                    QualifiedNameReference qualifiedNameReference = (QualifiedNameReference) entry.getValue();
+                    verifyTypeSignature(entry.getKey(), expectedType.getTypeSignature(), types.get(Symbol.fromQualifiedName(qualifiedNameReference.getName())).getTypeSignature());
+                    continue;
+                }
+                Type actualType = getExpressionTypes(session, metadata, sqlParser, types, entry.getValue()).get(entry.getValue());
+                verifyTypeSignature(entry.getKey(), expectedType.getTypeSignature(), actualType.getTypeSignature());
+            }
+        }
+
+        private static void verifyTypeSignature(Symbol symbol, TypeSignature expected, TypeSignature actual)
+        {
+            // NULL is of UNKNOWN, so UNKNOWN should be considered as a wildcard type, which matches all the other types
```

Comment:
This comment is a little misleading, but I'll let @haozhun suggest an alternative.